File: read_codebook.R

# This file contains functions that parse FEH codebook and return
# a data frame that contains a list of names of record fields, its
# types (scalar or array) and index ranges for arrays.


#' Returns a vector of field types
#' 
#' @param lines Character vector of codebook lines
#' @return A vector of field types
get_record_types <- function(lines)
{
  scalar_line_numbers <- grep("[[:space:]]*V-[[:space:]]*[[:alnum:]]+[[:space:]]+", lines)
  array_line_numbers <- grep("[[:space:]]*V-[[:space:]]*[[:alnum:]]+\\(", lines)
  rec_type <- character(length(lines))
  rec_type[scalar_line_numbers] <- "scalar"
  rec_type[array_line_numbers] <- "array"
  return(rec_type)
}

#' Parse codebook lines into a dataframe of names, types and array ranges
#' 
#' @param lines Character vector of codebook lines
#' @return A dataframe with variables: name, type, start, stop
parse_codebook_lines <- function(lines)
{
  per_names <- sub("[[:space:]]*V-[[:space:]]*([[:alnum:]]+)[[:space:](].*", "\\1", lines)
  rec_type <- get_record_types(lines)
  array_year_range <- sub("[[:space:]]*V-[[:space:]]*[[:alnum:]]+[[:space:]]+.*","", lines)
  array_year_range <- sub("[[:space:]]*V-[[:space:]]*[[:alnum:]]+\\(([[:digit:]]+\\-[[:digit:]]+)\\).*","\\1", array_year_range)
  array_year_start <- as.integer(sub("([[:digit:]]+)\\-.*","\\1", array_year_range))
  array_year_end   <- as.integer(sub("[[:digit:]]+\\-([[:digit:]]+).*","\\1", array_year_range))
  return(data.frame(name=per_names, type=rec_type, start=array_year_start, end=array_year_end))
}

#fam_df <- parse_codebook_lines(fam_rec_lines)

#' Parse person record codebook
#' 
#' @param fname The codebook file name
#' @return A dataframe with variables: name, type, start, stop
parse_person_codebook <- function(fname)
{
  codebook_text <- readLines(fname)
  
  fam_rec_start <- grep("[[:space:]]*RECORD- FAMILY", codebook_text)
  per_rec_start <- grep("[[:space:]]*RECORD- PERSON", codebook_text)
  endvars <- grep("[[:space:]]*ENDVARS", codebook_text)
  
  if(!(endvars[1] > fam_rec_start & endvars[1] < per_rec_start & endvars[2] > per_rec_start)) {
    stop("ERROR: Codebook invalid")
  }
  
  per_rec_lines <- codebook_text[(per_rec_start+1):(endvars[2]-1)]
  per_rec_lines <- grep("[[:space:]]*V-", per_rec_lines, value = T)
  
  return(parse_codebook_lines(per_rec_lines))
}

#' Parse family record codebook
#' 
#' @param fname The codebook file name
#' @return A dataframe with variables: name, type, start, stop
parse_family_codebook <- function(fname)
{
  codebook_text <- readLines(fname)
  
  fam_rec_start <- grep("[[:space:]]*RECORD- FAMILY", codebook_text)
  per_rec_start <- grep("[[:space:]]*RECORD- PERSON", codebook_text)
  endvars <- grep("[[:space:]]*ENDVARS", codebook_text)
  
  if(!(endvars[1] > fam_rec_start & endvars[1] < per_rec_start & endvars[2] > per_rec_start)) {
    stop("ERROR: Codebook invalid")
  }
  
  fam_rec_lines <- codebook_text[(fam_rec_start+1):(endvars[1]-1)]
  fam_rec_lines <- grep("[[:space:]]*V-", fam_rec_lines, value = T)

  return(parse_codebook_lines(fam_rec_lines))
}

File: Python/read_codebook.py

import re
import pandas as pd


def get_record_types(lines):
    
    scal_pat = '\s*V-\s+\w+\s+'
    arry_pat = '\s*V-\s+\S+\\('
    
    scalar_line_numbers = [i for i, line in enumerate(lines) if re.search(scal_pat, line)]
    array_line_numbers = [i for i, line in enumerate(lines) if re.search(arry_pat, line)]

    rec_type = [''] * len(lines)
    
    for index in scalar_line_numbers:
        rec_type[index] = 'scalar'
    for index in array_line_numbers:
        rec_type[index] = 'array'
    return rec_type


def parse_codebook_lines(lines):
    per_names = [re.sub(r'\s*V-\s*(\w+).*', r'\1', line).strip() for line in lines]
    rec_type = get_record_types(lines)  # Assuming get_record_types is defined elsewhere
    array_year_range = [re.sub(r'\s*V-\s*\w+\s+.*', '', line) for line in lines]
    array_year_range = [re.sub(r'\s*V-\s*\w+\((\d+\-\d+)\).*', r'\1', line) for line in array_year_range]
    array_year_start = [re.sub(r'(\d+)\-.*', r'\1', year_range).strip() for year_range in array_year_range]
    array_year_end = [re.sub(r'\d+\-(\d+).*', r'\1', year_range).strip() for year_range in array_year_range]
    
    return pd.DataFrame({'name': per_names, 'type': rec_type, 'start': array_year_start, 'end': array_year_end})


def parse_person_codebook(fname):

    with open(fname, 'r') as file:
        codebook_text = file.readlines()

    fam_rec_start = next(i for i, line in enumerate(codebook_text) if "RECORD- FAMILY" in line)
    per_rec_start = next(i for i, line in enumerate(codebook_text) if "RECORD- PERSON" in line)
    endvars = [i for i, line in enumerate(codebook_text) if "ENDVARS" in line]

    if not (endvars[0] > fam_rec_start and endvars[0] < per_rec_start and endvars[1] > per_rec_start):
        raise ValueError("ERROR: Codebook invalid")

    per_rec_lines = codebook_text[per_rec_start + 1:endvars[1]]
    per_rec_lines = [line for line in per_rec_lines if "V-" in line]

    return parse_codebook_lines(per_rec_lines)  # Assuming parse_codebook_lines is defined elsewhere

def df_to_col_names(df):
    rec_struct = df['name'].astype(str)
    
    arrays_df = df[df['type'] == 'array'][['name', 'start', 'end']]
    
    rec_struct = rec_struct.append(
        arrays_df.apply(
            lambda x: [x[0] + str(i) for i in range(int(x[1]), int(x[2]) + 1)],
            axis=1).explode(), 
        ignore_index=True)

    return rec_struct


def get_col_names(codebook):
    per_df = parse_person_codebook(codebook)
    col_names = df_to_col_names(per_df)
    return col_names

if __name__ == '__main__':

    fname = 'codebook_2087ds.sipp2006'

    df = get_col_names(fname)

    print(df)